d2/src/test/java/com/linkedin/d2/balancer/util/canary/TestBasicCanaryDistributionProviderImpl.java

```diff
@@ -183,28 +183,44 @@ public void testEdgeCases(D2CanaryDistributionStrategy strategy)
         CanaryDistributionProvider.Distribution.STABLE, "Invalid strategies should return stable");
   }
 
+  // Verify uniform distribution of the String built-in hashCode method, which theoretically should be guaranteed as it's used
+  // in Map with String keys. But we explicitly test it to verify its behavior.
   @Test
   public void testHashDistribution()
   {
     // verify hash results are almost uniformly distributed
-    int bucketCount = 100;
+    int bucketCount = 100; // since ramp is 100% based, we test with 100 buckets
     int attemptCount = 1000;
+    // count how many attempts are skewed
+    // (e.g: one bucket gets 13 will cause another bucket gets 7, the attempts skewed in this pair of buckets is ((13 - 10) + (10 - 7)) / 2 = 3
```

Comment:
I don't know if we should be dividing by two here. One bucket will receive 30% more, and one will receive 30% less, but the way you're doing the calculation would be against 20 attempts, and it would come out as 15%.

Reply:
The percentage calculated here is not how skewed a single bucket is, but how skewed the overall attempts are. If we want to distribute 1000 upstream client instances into 100 buckets, what we care about most is how many instances are mis-distributed (for example, expecting 400 instances fall into canary, but actually got 420, then the skewed instances are 20 / 400 = 0.05 = 5%, which means 5% of the clients will be mis-distributed).

So for this test, it first calculates how many attempts are mis-distributed (for example, expecting 10 attempts in each bucket, actually one bucket got 13, another got 7, then the mis-distributed attempts are 3, not 6), then divides it by the total attempts (1000) to get the percentage of mis-distributed attempts.

Reply:
modified the test to check the exact mis-distributed canary instances and added more comments to clarify what's tested. Moved the test to container since it uses linkedin-style host names for the test.